src/backend/catalog/heap.c

```diff
@@ -1409,8 +1414,26 @@ heap_create_with_catalog(const char *relname,
 		/*
 		 * We'll make an array over the composite type, too.  For largely
 		 * historical reasons, the array type's OID is assigned first.
+		 * 
+		 * For temp tables, we use temp OID assignment code here as well.
 		 */
-		new_array_oid = AssignTypeArrayOid();
+		if (use_temp_oid_buffer)
+		{
+			Relation	pg_type;
+
+			/* We should not be creating TSQL Temp Tables in pg_upgrade. */
+			Assert(!IsBinaryUpgrade);
+
+			pg_type = table_open(TypeRelationId, AccessShareLock);
+			new_array_oid = GetNewTempOidWithIndex_hook(pg_type, TypeOidIndexId, Anum_pg_type_oid);
+			
+			Assert(!OidIsValid(reltypeid));
+
+			reltypeid = GetNewTempOidWithIndex_hook(pg_type, TypeOidIndexId, Anum_pg_type_oid);
```

Comment:
we do not do this in the normal case. can you add a comment why?

Reply:
done